compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -1020,6 +1020,12 @@ object Types {
       case _ => this
     }
 
+    /** Widen singleton type modulo constant types */
+    final def widenNonConstant(implicit ctx: Context): Type = this match {
+      case _: ConstantType => this
+      case _ => widen
+    }
+
```

Comment:
It seems to me better to revert this change, as  (1) `tests/patmat/t11603.scala` is not a real use case and it's hard to imagine one; (2) it does not work if the scrutinee and pattern are both singletons (again why use patmat for such case).